src/ProducerEventProcessor/UploaderClientCaptureEventCollectorTest.cpp

```diff
@@ -0,0 +1,189 @@
+      CHECK(stop_called_);
+    }
+
+    CHECK(data_produce_thread_.joinable());
+    data_produce_thread_.join();
+
+    CHECK(data_upload_thread_.joinable());
+    data_upload_thread_.join();
+  }
+
+  UploaderClientCaptureEventCollector collector_;
+  size_t total_uploaded_data_bytes_ = 0;
+
+  absl::Mutex upload_finished_mutex_;
+  bool upload_finished_ ABSL_GUARDED_BY(upload_finished_mutex_) = false;
+  absl::Mutex early_stop_mutex_;
+  bool early_stop_ ABSL_GUARDED_BY(early_stop_mutex_) = false;
+
+ private:
+  // Keep producing capture events until stop is called or `event_count_` events are produced with a
+  // specified produce duration. The last produced event will always be the capture finish event.
+  void ProduceCaptureData() {
+    size_t added_event_count = 0;
+    bool need_check_early_stop = (early_stop_event_count_ >= 2);
+    bool stopped = false;
+    while (added_event_count < event_count_ - 1 && !stopped) {
+      if (need_check_early_stop && added_event_count + 2 >= early_stop_event_count_) {
+        absl::MutexLock lock{&early_stop_mutex_};
+        early_stop_ = true;
+        need_check_early_stop = false;
+      }
+
+      stop_called_mutex_.LockWhenWithTimeout(absl::Condition(&stop_called_),
+                                             absl::Milliseconds(produce_event_every_ms_));
+      collector_.AddEvent(CreateInternedStringCaptureEvent(kAnswerKey, kAnswerString));
+      stopped = stop_called_;
+      stop_called_mutex_.Unlock();
+
+      ++added_event_count;
+    }
+
+    collector_.AddEvent(CreateCaptureFinishedEvent());
+
+    Stop();
+  }
+
+  void UploadCaptureData() {
+    while (collector_.GetDataReadiness() != orbit_capture_uploader::DataReadiness::kEndOfData) {
+      std::this_thread::sleep_for(std::chrono::milliseconds(upload_event_every_ms_));
+
+      const std::vector<unsigned char>& upload_data_buffer = collector_.GetUploadDataBuffer();
+      total_uploaded_data_bytes_ += upload_data_buffer.size();
+      collector_.RefreshUploadDataBuffer();
+    }
+
+    absl::MutexLock lock{&upload_finished_mutex_};
+    upload_finished_ = true;
+  }
+
+  absl::Mutex stop_called_mutex_;
+  bool stop_called_ ABSL_GUARDED_BY(stop_called_mutex_) = false;
+
+  size_t event_count_ = 0;
+  size_t early_stop_event_count_;
+  uint64_t produce_event_every_ms_;
+  uint64_t upload_event_every_ms_;
+  std::thread data_upload_thread_;
+  std::thread data_produce_thread_;
+};
+
+}  // namespace
+
+TEST_F(UploaderClientCaptureEventCollectorTest, EarlyStop) {
+  constexpr size_t kEventCount = 10;
+  constexpr uint64_t kProduceEventEveryMs = 50;
+  constexpr uint64_t kUploadEventEveryMs = 5;
+  constexpr size_t kEarlyStopEventCount = 4;
+  Start(kEventCount, kProduceEventEveryMs, kUploadEventEveryMs, kEarlyStopEventCount);
+
+  // Stop producing data early
+  {
+    absl::MutexLock lock{&early_stop_mutex_};
+    early_stop_mutex_.Await(absl::Condition(&early_stop_));
+  }
+  Stop();
```

Comment:
I guess I don't understand this test. We wait until `early_stop_event_count` is reached, and then we call `UploaderClientCaptureEventCollectorTest::Stop`, which causes `ProduceCaptureData` to exit the loop and then call `UploaderClientCaptureEventCollectorTest::Stop` again. What is the difference with just calling `Start` with a lower `event_count`, e.g., `kEarlyStopEventCount` (+1?)? Or maybe you meant to call `collector_->Stop()` somewhere, maybe here?

Reply:
We will block `early_stop_mutex_` until the producer thread has produced `early_stop_event_count - 2` events.
Then we release the mutex and enable calling `Stop()`, which causes the `ProduceCaptureData` to exit the loop with producing one more `InternedStringCaptureEvent` and then produce a `CaptureFinishedEvent`.

> What is the difference with just calling Start with a lower event_count, e.g., kEarlyStopEventCount (+1?)? 

If we do not call `Stop()` in the test, the data_produce_thread will keep producing data until producing `kEventCount` event.

The test case without early stop is mainly to simulate the case that we take a capture until the game ends.

And the test case of early stop is to simulate the case that we call stop capture while the game is still running.

> Or maybe you meant to call collector_->Stop() somewhere, maybe here?

Now the `UploaderClientCaptureEventCollector::Stop()` is private, it will be called by the collector after processing the `CaptureFinishedEvent`.

Reply:
The purpose of using this `early_stop_event_count` is the determine the right time to release `early_stop_mutex_` and so we will able to call `Stop()` here. The previous version uses `std::this_thread::sleep_for` and so is not reliable. There is no way to know the exact count of produced events and the test might be flaky.

Reply:
> If we do not call Stop() in the test, the data_produce_thread will keep producing data until producing kEventCount event.
The test case without early stop is mainly to simulate the case that we take a capture until the game ends.
And the test case of early stop is to simulate the case that we call stop capture while the game is still running.

What I understand from this is that there's no practical difference in calling `Start` with a lower `event_count` from the point of view of the class under test. So the test only *looks* different, but it's not actually testing any different behavior of the class under test. As the machinery to do this is also somewhat involved (more mutexes, more fields, etc.), my opinion is that it should all just be removed (assuming I understood correctly!).

Either way, I'll leave it to you, or maybe @beckerhe picks up some detailed I missed.

Reply:
Update: Changed to remove the `EarlyStop` test case now and refactored the test a little bit.

Now we test two cases: start uploading after all data are produced, and start uploading right after start producing.